### PR TITLE
Domains: checkout multiple address formats

### PR DESCRIPTION
**Pull request by ramonjd**
This PR fixes the issue in PR #18515 that led to its revert (#19067)

### TL;DR

We are displaying a UK and EU style address fields depending on the country selected. Default is the current format (with state).

When no country is selected, we're hiding the address fields and disabling the button.

addresses-fix

### The fix
Removing all country codes for which the server returns a list of countries in [constants.js](https://github.com/Automattic/wp-calypso/blob/update/domain-checkout-multiple-address-formats/client/my-sites/domains/components/domain-form-fieldsets/constants.js).

Country codes removed: `BE`, `CN`, `ES`, `IT`

As a backup precautionary measure, we are also checking the store to see whether there any states for the country. If so, we make sure we display them.

We query the store state in `<DomainDetailsFormContainer />`  and not in `<RegionAddressFieldsets />` since `connect()` does a shallow comparison and won't trigger a re-render.

In the unlikely event that the error occurs again, we're also throwing an error (which is logged) and showing information in a notice, prompting the user to contact support.

screen shot 2017-10-26 at 12 43 00 pm

### Testing
After considering, and then deciding that creating automated integration tests for each country code might be going too far, I have tested submitted every country code in `constants.js` manually, including those that return states.

### A/B
[50/50 A/B test set up](https://github.com/Automattic/wp-calypso/blob/update/domain-checkout-multiple-address-formats/client/lib/abtest/active-tests.js#L102) that we can use in conjunction with the `calypso_contact_information_form_submit` and `calypso_contact_information_form_submit_details` events to test successful form submissions.

### Commentary
The reason why we had to revert I submit was due to the frontend not satisfying backend validation rules. By defining the state as an optional field, the frontend could not respond to an 'invalid state field' instruction from the backend, since the corresponding element didn't exist in the DOM.

This is another conversation, but it would be great to have a way to to communicate mandatory fields to the frontend for any checkout item/country variations ~~review the necessity of these rules. If ICANN doesn't require this information, [which it looks like it doesn't](https://tools.ietf.org/html/rfc5733#section-2.4), what are the advantages of requiring it on WordPress.com?~~.

@deBhal is working on a fix for `.ca` domains, the validation for which enforces a state regardless of the country.

### CODE REVIEW

**Comment by aidvu** [comment]
> This is another conversation, but it would be great to review the necessity of these rules. If ICANN doesn't require this information, which it looks like it doesn't, what are the advantages of requiring it on WordPress.com?

It's not only ICANN. We use external APIs for domains, and these might have their own validation in place.

**Comment by ramonjd** [comment]
> It's not only ICANN. We use external APIs for domains, and these might have their own validation in place.

Thanks @aidvu .

Would you have any tips on how get information on these validation requirements?

File: client/my-sites/checkout/checkout/domain-details-form.jsx [hunk 1]
```diff
@@ -429,7 +431,28 @@ export class DomainDetailsForm extends PureComponent {
 	focusFirstError() {
 		const firstErrorName = kebabCase( head( formState.getInvalidFields( this.state.form ) ).name );
 		const firstErrorRef = this.inputRefs[ firstErrorName ] || this.refs[ firstErrorName ];
-		firstErrorRef.focus();
+
+		try {
+			firstErrorRef.focus();
+		} catch ( err ) {
+			const noticeMessage = this.props.translate(
+				'There was a problem validating your contact details: %(firstErrorName) required. ' +
```
Comment (deBhal) [inline client/my-sites/checkout/checkout/domain-details-form.jsx]: Missing `s` from `%(firstErrorName)s`

File: client/my-sites/checkout/checkout/domain-details-form.jsx [hunk 2]
```diff
@@ -464,7 +431,43 @@ export class DomainDetailsForm extends PureComponent {
...
+			);
+			notices.error( noticeMessage );
+			throw new Error(
+				`Cannot focus() on invalid form element in domain details checkout form with name: '${ firstErrorName }'`
+			);
+		}
+	}
+
+	focusAddressField() {
+		const inputRef = this.inputRefs[ 'address-1' ] || null;
```
Comment (deBhal) [inline client/my-sites/checkout/checkout/domain-details-form.jsx]: It looks like we could drop the `|| null` here.

File: client/my-sites/checkout/checkout/domain-details-form.jsx [hunk 3]
```diff
@@ -429,7 +431,28 @@ export class DomainDetailsForm extends PureComponent {
 	focusFirstError() {
 		const firstErrorName = kebabCase( head( formState.getInvalidFields( this.state.form ) ).name );
 		const firstErrorRef = this.inputRefs[ firstErrorName ] || this.refs[ firstErrorName ];
-		firstErrorRef.focus();
+
+		try {
+			firstErrorRef.focus();
+		} catch ( err ) {
+			const noticeMessage = this.props.translate(
+				'There was a problem validating your contact details: "%(firstErrorName)s" required. ' +
```
Comment (akirk) [inline client/my-sites/checkout/checkout/domain-details-form.jsx]: Could we make `%(firstErrorName)s` a component as well? `{{firstErrorName/}}` and the component be something like `<strong>{firstErrorName}</strong>`

**Comment by deBhal** [comment]
I did some testing and checked against the values on the backend, and this works nicely :)

One thing I did turn up was ccTLDs. At the moment, `.CA` is marked as requiring a state field, so you can still trigger this error by trying to buy a .ca domain from Italy. I'm a bit on the fence about whether it's a blocker - I'll get a second opinion.

Apart from that, this looks good to go to me.  

For a different PR: It would be nice to get the data about domains that require state (and therefore should not use the UK fieldset) from the backend. I've just noticed that we've got  `countryStates` in the redux state (ugh, english). Are we able to use that to overrule the address format?

**Comment by klimeryk** [comment]
I love this idea... but hate the flow 😀 While testing this, I was a bit puzzled (as a user) that when I selected the country, a whole new section opened up _above_ the field I've just changed. Usually forms go top down, not top down, then up a bit ;) I'd love @mikeshelton1503's (or someone else from Cerulean) opinion on this, as any change on this form could have a high impact on conversion rates so we should not make them too lightly based on our assumptions and without design input.




Additionally, it seems we still need to handle a few cases better. Like, for example, new users tries to quickly go through this screen, without selecting a country (why would he? it doesn't seem like it's required). They'll get presented with a confusing error:
screen shot 2017-10-25 at 08 07 35

**Comment by ramonjd** [comment]
> Additionally, it seems we still need to handle a few cases better. Like, for example, new users tries to quickly go through this screen, without selecting a country (why would he? it doesn't seem like it's required). 

Thanks for catching this.

I also saw it this afternoon. This all stems from our attempt to focus on the first 'error'. I have a fix ready for it, but it's not pretty, which makes me think that all this thinking we're dedicating to form fields might be better spent simplifying the form and validation.

**Comment by ramonjd** [comment]
>  I've just noticed that we've got countryStates in the redux state (ugh, english). Are we able to use that to overrule the address format?

Looking in `state.countryStates.items` might be a good fallback, but as I mentioned before, I get the feeling we're adding fixes for fixes. 

As @deBhal wrote, it would be good to get data about domains that require state from the backend. In fact it would one day be nice to get all required fields from the backend so we can tell the user (and the clientside code) in advance what needs to be filled out instead of waiting for the button to be hit.

**Comment by mikeshelton1503** [comment]
Design 👍 

Works better now that the country doesn't switch order per @klimeryk's feedback.

File: client/my-sites/checkout/checkout/domain-details-form.jsx [hunk 4]
```diff
@@ -521,7 +537,7 @@ export class DomainDetailsForm extends PureComponent {
 		}
 	}
 
-	// We want to cache the functions to avoid triggering unecessary rerenders
+	// We want to cache the functions to avoid triggering unnecessary rerenders
```
Comment (klimeryk) [inline client/my-sites/checkout/checkout/domain-details-form.jsx]: 🤓

File: client/my-sites/checkout/checkout/notice-error-message.jsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ * @format
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+
+const NoticeErrorMessage = ( { message } ) => <strong>{ message }</strong>;
```
Comment (klimeryk) [inline client/my-sites/checkout/checkout/notice-error-message.jsx]: Is this really something we need as a component? The name is a bit confusing too, I was expecting it's somehow related to our `Notice` components.

Comment (ramonjd) [reply client/my-sites/checkout/checkout/notice-error-message.jsx]: I was working on the assumption that, if additions or changes were required to the translation variable, it would be easier via a component than updating a translation string (which has to then be re-translated/updated)  Maybe @akirk can confirm?

Comment (deBhal) [reply client/my-sites/checkout/checkout/notice-error-message.jsx]: Usually with something this simple we'd do it inline in the `translate()` arguments.
